packages/dashboard-frontend/src/containers/IdeLoader/index.tsx

```diff
@@ -0,0 +1,472 @@
+        nextStepCancellable = pseudoCancellable(this.startStep3(workspace));
+      }
+      this.toDispose.push({
+        dispose: () => {
+          nextStepCancellable.cancel();
+        },
+      });
+      const nextStep = await nextStepCancellable;
+      if (nextStep) {
+        this.switchToNextStep();
+      }
+    } catch (e) {
+      if (e instanceof Cancellation) {
+        // component updated, do nothing
+        return;
+      }
+      currentStep.hasError = true;
+      const lastError = common.helpers.errors.getMessage(e);
+      this.setState({
+        lastError,
+      });
+    }
+  }
+
+  private switchToNextStep(): void {
+    const { currentStepIndex } = this.state;
+    const currentStep = this.stepsList.get(currentStepIndex);
+
+    if (currentStep.hasNext() === false) {
+      return;
+    }
+
+    this.setState({
+      currentStepIndex: currentStep.next.index,
+    });
+  }
+
+  /**
+   * The resolved boolean indicates whether to go to the next step or not
+   */
+  private async runStep1(workspace: Workspace): Promise<boolean> {
+    if (workspace.isDeprecated) {
+      throw new Error(`The workspace is deprecated. Convert the workspace and try again.`);
+    }
+
+    if (workspace.status === DevWorkspaceStatus.TERMINATING) {
+      throw new Error(`The workspace is terminating and cannot be open.`);
+    }
+
+    // if stopping / failing
+    if (
+      workspace.status === WorkspaceStatus.STOPPING ||
+      workspace.status === DevWorkspaceStatus.STOPPING ||
+      workspace.status === DevWorkspaceStatus.FAILING
+    ) {
+      try {
+        await new Promise<void>((resolve, reject) => {
+          const timeoutId = window.setTimeout(() => {
+            reject();
+          }, TIMEOUT_TO_STOP_SEC * 1000);
+
+          this.toDispose.push({
+            dispose: () => {
+              window.clearTimeout(timeoutId);
+              resolve();
+            },
+          });
+        });
+
+        // do not switch to the next step
+        return false;
+      } catch (e) {
+        throw new Error(
+          `The workspace status remains "${workspace.status}" in the last ${TIMEOUT_TO_STOP_SEC} seconds.`,
+        );
+      }
+    }
+
+    // switch to the next step
+    await delay(MIN_STEP_DURATION_MS);
+    return true;
+  }
+
+  /**
+   * The resolved boolean indicates whether to go to the next step or not
+   */
+  private async runStep2(workspace: Workspace): Promise<boolean> {
+    if (
+      workspace.status === DevWorkspaceStatus.TERMINATING ||
+      workspace.status === WorkspaceStatus.STOPPING ||
+      workspace.status === DevWorkspaceStatus.STOPPING ||
+      workspace.status === DevWorkspaceStatus.FAILING ||
+      (this.state.shouldStart === false &&
+        (workspace.status === WorkspaceStatus.STOPPED ||
+          workspace.status === WorkspaceStatus.ERROR ||
+          workspace.status === DevWorkspaceStatus.STOPPED ||
+          workspace.status === DevWorkspaceStatus.FAILED))
```

Comment:
I see a lot of comparing for ```workspace.status```, like this:
```
      workspace.status === DevWorkspaceStatus.TERMINATING ||
      workspace.status === DevWorkspaceStatus.STOPPING ||
      workspace.status === DevWorkspaceStatus.FAILING
...
```
I think it will be useful to use a separate function for comparing
```
  checkStatus(workspace: Workspace, ...args: DevWorkspaceStatus[]): boolean {
    for (const status of args) {
      if (status === workspace.status) {
        return true;
      }
    }
    return false;
  }
```

WDYT?

Reply:
that makes sense, thank you!